lab/src/mapreduce/schedule.go

```diff
@@ -30,5 +33,42 @@ func schedule(jobName string, mapFiles []string, nReduce int, phase jobPhase, re
 	//
 	// Your code here (Part III, Part IV).
 	//
+
+	var workersPool sync.WaitGroup
+	var taskLock sync.Mutex
+
+	undoTask := ntasks
+	taskRemain := make(chan int, ntasks)
+	for taskNum := 0; taskNum < ntasks; taskNum++ {
+		taskRemain <- taskNum
+	}
+
+	for currentTaskNum := range taskRemain {
+		worker := <- registerChan
+		workersPool.Add(1)
+		go func(taskNum int, taskRemain chan int) {
+			ok := call(worker, "Worker.DoTask", DoTaskArgs{jobName, mapFiles[taskNum], phase, taskNum, n_other}, nil)
+			
+			defer workersPool.Done()
+			// register back to master
+			go func() {registerChan <- worker}()
+
+			if ok == false {
+				taskRemain <- taskNum
+			} else {
+				taskLock.Lock()
```

Comment:
Do we need to lock `undoTask` if there's always only one goroutine running at the same time?

Reply:
@Shauro waitGroup allow multiple goroutine run, so there will be a race condition on undoTask I think :O